test/Hedgehog/Extras/Test/TestWatchdogSpec.hs

```diff
@@ -0,0 +1,137 @@
+{-# LANGUAGE NumericUnderscores #-}
+module Hedgehog.Extras.Test.TestWatchdogSpec where
+
+import           Control.Concurrent
+import           Control.Monad
+import           Control.Monad.IO.Class
+import           Control.Monad.Trans.Except (runExceptT)
+import           Control.Monad.Trans.Resource (ResourceT, runResourceT)
+import           Control.Monad.Trans.Writer.Lazy (runWriterT)
+import           Data.Function
+import           Data.List (isPrefixOf)
+import           Data.Time.Clock as D
+import           GHC.Conc
+import           GHC.Stack
+import           Hedgehog (Property, (===))
+import qualified Hedgehog as H
+import qualified Hedgehog.Extras as H
+import qualified Hedgehog.Extras.Stock as H
+import           Hedgehog.Extras.Test.TestWatchdog
+import qualified Hedgehog.Internal.Property as H
+import           Prelude
+import qualified System.Process as P
+
+-- | Check that watchdog kills test case which waits without an end
+hprop_check_watchdog_kills_hanged_thread :: Property
+hprop_check_watchdog_kills_hanged_thread = H.propertyOnce $ do
+  let watchdogCfg = WatchdogConfig 1
+  childTid <- H.newEmptyMVar
+  tripwire <- H.makeTripwire
+
+  (result, _) <- spawnTestT $ runWithWatchdog_ watchdogCfg $ do
+    liftIO $ myThreadId >>= H.putMVar childTid
+    -- simulate thread hang
+    void $ H.threadDelay 3_000_000
+    H.triggerTripwire tripwire
+
+  _ <- case result of
+    Right () -> do
+      H.note_ "Expected failure instead of Right"
+      H.failure
+    Left (H.Failure _ msg _) -> do
+      -- check we've failed because of watchdog
+      _header:exception:_ <- pure $ lines msg
+      H.assertWith exception $
+        isPrefixOf "WatchdogException: "
+
+  -- make sure that we  didn't trigger the tripwire
+  H.checkTripwire tripwire
+  childStatus <- liftIO $ H.readMVar childTid >>= threadStatus
+  childStatus === ThreadFinished
+
+-- | Check that watchdog kills test case which spawns:
+--  - a process waiting forever
+--  - a child threads waiting
+hprop_check_watchdog_kills_hanged_thread_with_its_children :: Property
+hprop_check_watchdog_kills_hanged_thread_with_its_children = H.propertyOnce $ do
+  let watchdogCfg = WatchdogConfig 1
+  childTid <- H.newEmptyMVar
+  grandChildTid1 <- H.newEmptyMVar
+  grandChildTid2 <- H.newEmptyMVar
+  procHandle <- H.newEmptyMVar
+  childTripwire <- H.makeTripwire
+  grandChildTripwire1 <- H.makeTripwire
+  grandChildTripwire2 <- H.makeTripwire
+
+  (result, _) <- spawnTestT $ runWithWatchdog_ watchdogCfg $ do
+    liftIO $ myThreadId >>= H.putMVar childTid
```

Comment:
We should assert that this was executed using a tripwire.

Reply:
This implicitly is done by `MVar`. It's checked in line 94:
```haskell
  H.byDeadlineM 0.2 deadline "grandChildStatus1" $ do
    grandChildStatus1 <- liftIO $ H.readMVar grandChildTid1 >>= threadStatus
    grandChildStatus1 === ThreadFinished
```
so if line 67 was never reached, `MVar` will be empty, `readMVar` will block, but only until `deadline` passes, so it will report a failure